### ISSUE TITLE
How to deserialize toml config to a vector of enum struct varient?

### ISSUE THREAD

**Issue by HongbinZhou** [alexcrichton/toml-rs#222]
**TLDR:** How can I deserialize the following toml configuration str to Rust struct `Vec<Layer>`?

Deserialize config:

    [input]
    ndim = 2
    dims = [10, 100]

    [dense]
    outputs = 128

    [lstm]
    outputs =64
    return_seq = true

to a vector of `Layer`:

    [
       Input(ndim: 2, dims: [10, 100]),
       Dense(outputs: 128),
       Lstm(outputs: 64, return_seq: true)
    ]

while `Layer` is:

    enum Layer {
        Input { ndim: u32, dims: Vec<u32> },
        Lstm { outputs: u32, return_seq: bool },
        Dense { outputs: u32 },
    }

**My issue detailed description:**

I have a toml config file for a deep learning task similar as above shown, it
has one or more layers, each layer has its own key-value parameters. The layer
order matters.

I tried 3 ways but all failed.

**Try 1:** https://play.rust-lang.org/?gist=86b367e3148400f1090451c2fb1d7860&version=nightly

Using the following structures can deserilize the above config file but it
doesn't work for other configrations.

    #[derive(Serialize, Deserialize, Debug)]
    struct FixedLayersOnlyWorkForThisConfig {
        input: Input,
        dense: Dense,
        lstm: Lstm
    }
    #[derive(Serialize, Deserialize, Debug)]
    struct Layer {
        input: Input,
        lstm: Lstm,
    }
    #[derive(Serialize, Deserialize, Debug)]
    struct Input {
        ndim: u32,
        dims: Vec<u32>,
    }
    #[derive(Serialize, Deserialize, Debug)]
    struct Lstm {
        outputs: u32,
    }

For example it can't deserialize below config anymore:

    [input]
    ndim = 2
    dims = [10, 100]

    [dense]
    outputs = 128

**Try 2:** https://play.rust-lang.org/?gist=d88adf7a92c7063f955c59c85aa7149c&version=nightly

Using `toml::Value` `FromStr` can parse my config to toml internal structure:

    Table({
           "dense": Table({"outputs": Integer(128)}),
           "input": Table({"dims": Array([Integer(10), Integer(100)]), "ndim": Integer(2)}),
           "lstm" : Table({"outputs": Integer(64), "return_seq": Boolean(true)})
         })

But as toml used `BTreeMap` instead of `HashMap` for `Table`, **the order of layers
are not kept anymore which is not acceptable.**

**Try 3:** https://play.rust-lang.org/?gist=7a8cba0690aa850e189dc630652e742c&version=nightly

Then I tried toml deserialize, I expected it can be deserialized to type: `Vec<Layer>`
which looks like like:

    [
       Input(ndim: 2, dims: [10, 100]),
       Dense(outputs: 128),
       Lstm(outputs: 64, return_seq: true)
    ]

However the following function failed with error:

    Err(Error { inner: ErrorInner {kind: Custom, line: None, col: 0,
                                   message: "invalid type: map, expected a sequence", key: [] }})

Do I use toml deserialize wrong? Do I need to manually `impl Deserilize` for `Layer`?

Thanks a lot!

**Comment by alexcrichton** [comment]
Thanks for the report! This may not be the easiest thing to do unfortunately though. You will probably need to implement your own version of the `Deserialize` trait as currently a map can only decode into a map-like structure automatically. With a custom `Deserialize` implementation though you can perhaps translate it to a vector of `Layer` instances. Would that work for you?

**Comment by HongbinZhou** [comment]
Thanks for the reply Alex! I tried to impl `Deserialize` myself for a vector of `Layer` instances. My idea is as each `Layer` is a map in toml, I need to call `deserialize_map()`, then use a `LayerVisitor` to translate that map to a `Layer` instance and pushed to a `Vec<Layer>`. 

The code is here: https://play.rust-lang.org/?gist=614d5057e4d4bb5bcfed7301c4304525&version=nightly , but it build failed `"error[E0117]: only traits defined in the current crate can be implemented for arbitrary types". `  

(I hacked by changing return value form `Vec<Layer>` to `Layer` and return a dumb single layer in `deserialize()`, it can build and the printed vector value is what I wanted: https://play.rust-lang.org/?gist=3c876833cf2f47bc879290fb96a30388&version=nightly)

Do you have any further suggestions? thank you for your help!

**Comment by alexcrichton** [comment]
Ah yeah you'll need `Network` to be a custom type if you want to implement `Deserialize` for it. You can also shorten it a bit with `#[derive(Deserialize)]` on each layer like so -- https://play.rust-lang.org/?gist=78616877280c686137bd5d0a63811623&version=nightly.

I *think* you could get extra fancy and actually hook it up to automatically infer key names from enum variants, but that's a bit outside my serde-fu.

**Comment by HongbinZhou** [comment]
Brilliant!!

Indeed I should make `Network` as a custom type instead of a type alias. Your reminder of automatically infer enum variants is also helpful! Using serde untagged enum can get rid of the redundant like `Input(InputLayer)`, see: https://play.rust-lang.org/?gist=7b2077ffa87292bb24843509b7cf2c5d&version=nightly

Thank you Alex for your kind and quick help! I learned a lot in this case. I put some notes here for my future reference:

- There is no default API to deserialize several toml tables to a vector of tables. We can define a custom `Deserialize`.

- For each toml table we can `#[derive(Deserialize)]`. And using `#[serde(untagged)]` will deserialize the toml table values only. We can validate the toml table name in our custom `Deserialize`.

- Must impl `Deserialize` to a new type (via struct or enum) not for type alias (via `type newtype = ...`). This is because `Deserialize` is not a local trait, and "only traits defined in the current crate can be
implemented for arbitrary types".

I pasted the code here in case someone may also find it useful too.

```
extern crate serde;
#[macro_use]
extern crate serde_derive;

extern crate toml;

use std::fmt;
use serde::de::{self, Deserialize, MapAccess, Error};


const LAYERS_TOML: &'static str = r#"

[input]
ndim = 2
dims = [10, 100]

[dense]
outputs = 128

[lstm]
outputs =64
return_seq = true
"#;

#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum Layer {
    Lstm { outputs: u32, return_seq: bool },
    Dense { outputs: u32 },
    Input { ndim: u32, dims: Vec<u32> },
}

#[derive(Debug)]
struct Network {
    layers: Vec<Layer>,
}

impl Network {
    fn get_layer_num(&self) -> usize {
        self.layers.len()
    }
}

impl<'de> Deserialize<'de> for Network {
    fn deserialize<D>(deserializer: D) -> Result<Network, D::Error>
        where D: de::Deserializer<'de>
    {
        struct LayerVisitor;

        impl<'de> de::Visitor<'de> for LayerVisitor {
            type Value = Network;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("any valid layer")
            }

            fn visit_map<V>(self, mut visitor: V) -> Result<Network, V::Error>
                where V: MapAccess<'de>
            {
                let mut v = Vec::new();

                while let Some(key) = visitor.next_key::<String>()? {
                    match key.as_ref() {
                        "input" => v.push(visitor.next_value()?),
                        "dense" => v.push(visitor.next_value()?),
                        "lstm" => v.push(visitor.next_value()?),
                        _ => return Err(V::Error::custom("invalid layer name")),
                    }
                }

                Ok(Network { layers: v })
            }
        }

        deserializer.deserialize_map(LayerVisitor)
    }
}

fn test() {
    let layers: Result<Network, ::toml::de::Error> = toml::from_str(&LAYERS_TOML);
    if let Ok(layers) = layers {
        println!("layers de from toml:\n {:#?}", layers);
        println!("layers number: {}", layers.get_layer_num());

    } else {
        println!("layers de from toml failed:\n {:#?}", layers);
    }

}

fn main() {
    test()
}
```
